chemprop/cli/train.py

```diff
@@ -542,170 +531,220 @@ def main(args):
+            else:
+                mp_block = mp_cls(
+                    train_dset.featurizer.atom_fdim,
+                    train_dset.featurizer.bond_fdim,
+                    d_h=args.message_hidden_dim,
+                    bias=args.message_bias,
+                    depth=args.depth,
+                    undirected=args.undirected,
+                    dropout=args.dropout,
+                    activation=args.activation,
+                )
+            agg = Factory.build(AggregationRegistry[args.aggregation], norm=args.aggregation_norm)
+            predictor_cls = PredictorRegistry[args.task_type]
+
+            if args.loss_function is not None:
+                criterion = Factory.build(
+                    LossFunctionRegistry[args.loss_function],
+                    v_kl=args.v_kl,
+                    threshold=args.threshold,
+                    eps=args.eps,
+                )
+            else:
+                logger.info(
+                    f"No loss function specified, will use class default: {predictor_cls._default_criterion}"
+                )
+                criterion = predictor_cls._default_criterion
+
+            d_xf = (
+                sum(dset.d_xf for dset in train_dset.datasets)
+                if multicomponent
+                else train_dset.d_xf
+            )
+            predictor_ffn = Factory.build(
+                predictor_cls,
+                input_dim=mp_block.output_dim + d_xf,
+                n_tasks=train_dset.datasets[0].Y.shape[1]
+                if multicomponent
+                else train_dset.Y.shape[1],
+                hidden_dim=args.ffn_hidden_dim,
+                n_layers=args.ffn_num_layers,
+                dropout=args.dropout,
+                activation=args.activation,
+                criterion=criterion,
+                n_classes=args.multiclass_num_classes,
+                spectral_activation=args.spectral_activation,
+            )
+
+            if isinstance(predictor_ffn, RegressionFFN):
+                scaler = train_dset.normalize_targets()
+                val_dset.normalize_targets(scaler)
+                logger.info(f"Train data: loc = {scaler.mean_}, scale = {scaler.scale_}")
+            else:
+                scaler = None
+
+            train_loader = data.MolGraphDataLoader(train_dset, args.batch_size, args.num_workers)
+            val_loader = data.MolGraphDataLoader(
+                val_dset, args.batch_size, args.num_workers, shuffle=False
+            )
+
+            if multicomponent:
+                if len(test_data[0]) > 0:
+                    test_dsets = [make_dataset(data, args.rxn_mode) for data in test_data[fold_idx]]
+                    test_dset = data.MulticomponentDataset(test_dsets)
+                    test_loader = data.MolGraphDataLoader(
+                        test_dset, args.batch_size, args.num_workers, shuffle=False
+                    )
+                else:
+                    test_loader = None
```

Comment:
This can be moved outside of the `if multicomponent:` block right? (and removed from both places in here).

Reply:
Thanks for the suggestion! I move this outside of the `if multicomponent:` block.